crates/primitives/src/snapshot/segment.rs

```diff
@@ -176,7 +180,69 @@ impl SegmentHeader {
 
     /// Returns the first transaction number of the segment.
     pub fn tx_start(&self) -> TxNumber {
-        *self.tx_range.start()
+        *self.tx_range.as_ref().expect("should exist").start()
+    }
+
+    /// Returns the last transaction number of the segment.
+    pub fn tx_end(&self) -> TxNumber {
+        *self.tx_range.as_ref().expect("should exist").end()
+    }
+
+    /// Number of transactions.
+    pub fn tx_len(&self) -> u64 {
+        self.tx_range.as_ref().expect("should exist").end() + 1 -
+            self.tx_range.as_ref().expect("should exist").start()
+    }
+
+    /// Number of blocks.
+    pub fn block_len(&self) -> u64 {
+        self.block_range.end() + 1 - self.block_range.start()
+    }
+
+    /// Increments block end range depending on segment
+    pub fn increment_block(&mut self) {
+        self.block_range = *self.block_range.start()..=*self.block_range.end() + 1;
+    }
+
+    /// Increments tx end range depending on segment
+    pub fn increment_tx(&mut self) {
+        match self.segment {
+            SnapshotSegment::Headers => (),
+            SnapshotSegment::Transactions | SnapshotSegment::Receipts => {
+                if let Some(tx_range) = &mut self.tx_range {
+                    let start = if *tx_range.start() == 0 { 1 } else { *tx_range.start() };
+                    *tx_range = start..=*tx_range.end() + 1;
+                } else {
+                    self.tx_range = Some(1..=1);
+                }
+            }
+        }
+    }
+
+    /// Removes `num` elements from end of tx range.
```

Comment:
this sometimes also deals with block_range, would be nice to reflect this in the comment